### ISSUE TITLE
Dates

### ISSUE THREAD

**Issue by matthieugomez** [JuliaData/CategoricalArrays.jl#344]
Shouldn't CategoricalArrays work with Dates?
```
using Dates, CategoricalArrays
categorical([Date(01, 01, 01)])
ERROR: ArgumentError: CategoricalArray only supports AbstractString, AbstractChar and Number element types (got element type Date)
```

**Comment by nalimilan** [comment]
What's the use case? You're unlikely to store dates in a custom order which would differ from the time scale.

**Comment by eloualiche** [comment]
It's useful as a generic conversion when using dates as fixed effects. Right now the workaround is to convert them to strings before converting them to a categorical. This would be a lot more convenient. 
Thank you!

**Comment by nalimilan** [comment]
What modeling package are you using? How does it treat dates by default?

**Comment by eloualiche** [comment]
I use [FixedEffectModels](https://github.com/FixedEffects/FixedEffectModels.jl). I don't think it does anything specific with dates, meaning since dates can't be converted to categorical, they can't be used as fixed effects.

R for example allows to set pretty much anything as a factor, I think this could be a good feature to add.

**Comment by matthieugomez** [comment]
@nalimilan is right, dates are treated as categorical variables by default.

**Comment by nalimilan** [comment]
> I use [FixedEffectModels](https://github.com/FixedEffects/FixedEffectModels.jl). I don't think it does anything specific with dates, meaning since dates can't be converted to categorical, they can't be used as fixed effects.

Can you confirm whether they are treated as categorical by default as @matthieugomez said? That sounds like a reasonable fallback for non-numeric types.

> R for example allows to set pretty much anything as a factor, I think this could be a good feature to add.

On the contrary, R only allows the equivalent of `CategoricalArray{String}`. CategoricalArrays is much more flexible as it also allows numbers, but we had to drop support for arbitrary types to reduce compilation latency (https://github.com/JuliaData/CategoricalArrays.jl/issues/177). More types can easily be added, but I'd rather limit the list to the most useful types or it would end up being hard to follow.

**Comment by eloualiche** [comment]
I am confused about your statement about `R`. Can you explain what you mean?
The following commands seem to work:
```R
as.factor(c(1,2))
as.factor(c(as.Date("2020-04-15"), as.Date("2020-04-16")))
```

Your point is well taken about speed. And I agree that it has made a major difference in the performance when I use categorical.

I have to confirm with @matthieugomez. I think dates do work as categorical in fixedeffetmodels. So this was a blunder on my end (and I cannot find  the example that triggered this discussion in the first place).
I think it's ok to  close this and I will reopen it if I find the example.

**Comment by nalimilan** [comment]
Thanks for confirming.

> I am confused about your statement about `R`. Can you explain what you mean?
> The following commands seem to work:
> 
> ```r
> as.factor(c(1,2))
> as.factor(c(as.Date("2020-04-15"), as.Date("2020-04-16")))
> ```

R is just automatically converting to string under the hood. This gives possibly misleading results like this:
```R
> as.factor(c(as.Date("2020-04-15"), as.Date("2020-04-16"))) == as.Date("2020-04-15")
[1] FALSE FALSE
Warning message:
Incompatible methods ("Ops.factor", "Ops.Date") for "==" 
```
